### ISSUE TITLE
teeu安了scorecardpy 重新build 执行脚本还是找不到python模块

### ISSUE THREAD

**Issue by Jymjiaym** [secretflow/secretflow#1125]
图一是用python 命令行方式import没问题 ；但是下边用occlum方式执行，报错找不到scorecardpy模块；图二是pip成功后的截图,帮忙看下这个问题。
![1](https://github.com/secretflow/secretflow/assets/93430715/5ef49847-491c-4b6d-8f2a-3b80f33039d7)

![2](https://github.com/secretflow/secretflow/assets/93430715/2a6fbaa8-ea08-475f-91ad-6b3c405a67d3)

**Comment by zhouaihui** [comment]
在occlum上安装python库的方式参见

https://www.secretflow.org.cn/docs/secretflow/latest/zh-Hans/developer/design/teeu#how-to-install-and-use-new-python-libraries-in-occlum

**Comment by Jymjiaym** [comment]
> 在occlum上安装python库的方式参见
> 
> https://www.secretflow.org.cn/docs/secretflow/latest/zh-Hans/developer/design/teeu#how-to-install-and-use-new-python-libraries-in-occlum
我按照上边链接重新install后 重新build后 报错如下截图
![3](https://github.com/secretflow/secretflow/assets/93430715/08d85898-b62e-4170-835c-1d94d533423f)

**Comment by Jymjiaym** [comment]
> 在occlum上安装python库的方式参见
> 
> https://www.secretflow.org.cn/docs/secretflow/latest/zh-Hans/developer/design/teeu#how-to-install-and-use-new-python-libraries-in-occlum

执行脚本报错如下截图：
![Uploading 5.jpg…]()

**Comment by Jymjiaym** [comment]
> 在occlum上安装python库的方式参见
> 
> https://www.secretflow.org.cn/docs/secretflow/latest/zh-Hans/developer/design/teeu#how-to-install-and-use-new-python-libraries-in-occlum

执行脚本报错
![5](https://github.com/secretflow/secretflow/assets/93430715/812b584d-c54c-477a-ac26-d6e0c1a0f7c8)

**Comment by zhouaihui** [comment]
麻烦贴一下 scorecard.py的内容。

**Comment by Jymjiaym** [comment]
> 麻烦贴一下 scorecard.py的内容。

#1、两方数据合并部分
#2、评分卡建模流程（以下部分）
import scorecardpy as sc
import secretflow as sf
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score, roc_curve, auc
from sklearn.model_selection import GridSearchCV as gscv
from sklearn.neighbors import KNeighborsClassifier
import toad
#import math
#import glob
#import matplotlib.pyplot as plt
#import seaborn as sns
#import xgboost as xgb
#from toad.plot import bin_plot
import joblib
def train(x,y):

 #df为两方已整合在一起的数据（含label：SeriousDlqin2yrs）
 df = pd.merge(x,y,on= 'id')
 #df_test = pd.read_csv('cs-test.csv') #此数据集没有y值
 df=df.drop('id',axis=1)
 #缺失值中位数填充
 df=df.fillna(df.median())
 #数据集拆分
 df_train_test = df.head(int(df.shape[0]*0.75))
 df_oot = df.tail(int(df.shape[0]*0.25))
 df_train, df_test = sc.split_df(df_train_test,'SeriousDlqin2yrs',ratio=0.7).values()
 print("df_train.shape: ",df_train.shape)
 #缺失率（上一步已填充）、同值率筛选，iv初筛（可设0）
 df_s = sc.var_filter(df_train, y="SeriousDlqin2yrs",iv_limit=0.02,missing_limit=0.98, identical_limit=0.98)
 print("df_s.shape: ",df_s.shape)
 #基于tree的分箱
 bins = sc.woebin(df_s, method='tree',count_distr_limit=0.04,bin_num_limit=5,y="SeriousDlqin2yrs")
 print("bins: ",bins['age'])
 #分箱调整
 breaks_adj = {'age': [35,55,65]}
 bins_adj = sc.woebin(df_s, breaks_list=breaks_adj,method='tree',count_distr_limit=0.04,bin_num_limit=5,y="SeriousDlqin2yrs")
 print("bins_adj: ",bins_adj['age'])
 #iv值筛选
 fea_delete_iv=list({k:v['total_iv'][0] for k,v in bins_adj.items() if v['total_iv'][0]<=0.1}.keys())
 for i in fea_delete_iv:
     del bins_adj[i]
 fea_after_iv = list(bins_adj.keys())
 print("fea_after_iv: ",fea_after_iv)
 fea_y=fea_after_iv
 fea_y.append('SeriousDlqin2yrs')
 print("fea_y: ",fea_y)
 #相关性筛选
 corr_selected, dropped = toad.selection.select(df_train[fea_y],target = 'SeriousDlqin2yrs', empty =0, iv =0, corr = 0.7, return_drop=True)
 fea_selected=list(corr_selected.columns)
 #woe transfer
 train_woe = sc.woebin_ply(df_train[fea_selected], bins_adj)
 test_woe = sc.woebin_ply(df_test[fea_selected], bins_adj)
 #特征psi检验
 print(sc.perf_psi(
   score = {'train':pd.DataFrame(round(train_woe['age_woe'],4)), 'test':pd.DataFrame(round(test_woe['age_woe'],4))},
   label = {'train':train_woe['SeriousDlqin2yrs'], 'test':test_woe['SeriousDlqin2yrs']},
   show_plot=False
 ))
 print(sc.perf_psi(
   score = {'train':pd.DataFrame(round(train_woe['RevolvingUtilizationOfUnsecuredLines_woe'],4)), 'test':pd.DataFrame(round(test_woe['RevolvingUtilizationOfUnsecuredLines_woe'],4))},
   label = {'train':train_woe['SeriousDlqin2yrs'], 'test':test_woe['SeriousDlqin2yrs']},
   show_plot=False
 ))
 print(sc.perf_psi(
   score = {'train':pd.DataFrame(round(train_woe['NumberOfTimes90DaysLate_woe'],4)), 'test':pd.DataFrame(round(test_woe['NumberOfTimes90DaysLate_woe'],4))},
   label = {'train':train_woe['SeriousDlqin2yrs'], 'test':test_woe['SeriousDlqin2yrs']},
   show_plot=False
 ))
 #模型训练
 y_train = train_woe.loc[:,'SeriousDlqin2yrs']
 x_train = train_woe.loc[:,train_woe.columns != 'SeriousDlqin2yrs']
 y_test = test_woe.loc[:,'SeriousDlqin2yrs']
 x_test = test_woe.loc[:,train_woe.columns != 'SeriousDlqin2yrs']
 from sklearn.linear_model import LogisticRegression
 lr = LogisticRegression(penalty='l2', C=0.9, solver='saga', n_jobs=-1)
 lr.fit(x_train, y_train)
 
 train_pred = lr.predict_proba(x_train)[:,1]
 test_pred = lr.predict_proba(x_test)[:,1]
 test_predict=lr.predict(x_test)
 print('perf_eva train: ',sc.perf_eva(y_train, train_pred, title = "train",show_plot=False))
 print('perf_eva test: ',sc.perf_eva(y_test, test_pred, title = "test",show_plot=False))
 #评分卡转换
 # score ------
 card = sc.scorecard(bins_adj, lr, x_train.columns)
 # credit score
 train_score = sc.scorecard_ply(df_train, card, print_step=0)
 test_score = sc.scorecard_ply(df_test, card, print_step=0)
 #转换后检验
 #转换后的ks
 sc.perf_eva(y_train, np.array(train_score.iloc[:,0]), title = "train",show_plot=False)
 sc.perf_eva(y_test, np.array(test_score.iloc[:,0]), title = "test",show_plot=False)
 #评分分布&psi
 sc.perf_psi(
   score = {'train':train_score, 'test':test_score},
   label = {'train':y_train, 'test':y_test},
   show_plot=False
 )
 #提升度
 df_result=pd.DataFrame({"label":np.array(y_test),"predict":test_predict,"pos_probablity":test_pred,"score":np.array(test_score.iloc[:,0])})
 df_result['score_group'] = pd.qcut(df_result['score'],8,duplicates='drop')
 #保存df_result
 print("df_result: ",df_result.shape)
 df_result_bin = pd.crosstab(df_result['score_group'],df_result['label'])
 df_result_bin.insert(0, 'cut_number', range(0, len(df_result_bin))) # 添加序列号 方便排序
 df_result_bin['total'] = df_result_bin[0]+df_result_bin[1]
 df_result_bin['cum_bad'] = df_result_bin[1].cumsum()
 df_result_bin['bad_by_model'] = df_result_bin[1]/df_result_bin[1].sum()
 df_result_bin['bad_by_random'] = df_result_bin['total']/df_result_bin['total'].sum() # random 分子分母同时乘以bad_rate 固定
 df_result_bin['cumbad_by_model'] = df_result_bin['cum_bad']/df_result_bin[1].sum()
 df_result_bin['cumbad_by_random'] = df_result_bin['total'].cumsum()/df_result_bin['total'].sum()
 df_result_bin['lift'] = df_result_bin['cumbad_by_model']/df_result_bin['cumbad_by_random']
 #保存df_result_bin
 print("df_result_bin: ",df_result_bin)
 joblib.dump(card,'./scorecard.pkl')
def saveFile(name,value):
  filednames=['result_type','value']
  with open('./result.csv','w',newline='') as csvfile:
   writer=csv.DictWriter(csvfile,filednames=fieldnames)
   writer.writeheader()
   writer.writerow({'result_type':name,'value':value})
def get_data(path):
 data=pd.read_csv(path)
 return data
data1=get_data('./cs-training1.csv')
data2=get_data('./cs-training2.csv')
train(data1,data2)

**Comment by Jymjiaym** [comment]
`import scorecardpy as sc
import secretflow as sf
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score, roc_curve, auc
from sklearn.model_selection import GridSearchCV as gscv
from sklearn.neighbors import KNeighborsClassifier
import toad
import joblib

def train(x, y):
    # df为两方已整合在一起的数据（含label：SeriousDlqin2yrs）
    df = pd.merge(x, y, on='id')
    df = df.drop('id', axis=1)
    
    # 缺失值中位数填充
    df = df.fillna(df.median())
    
    # 数据集拆分
    df_train_test = df.head(int(df.shape[0] * 0.75))
    df_oot = df.tail(int(df.shape[0] * 0.25))
    df_train, df_test = sc.split_df(df_train_test, 'SeriousDlqin2yrs', ratio=0.7).values()
    
    print("df_train.shape: ", df_train.shape)
    
    # 缺失率（上一步已填充）、同值率筛选，iv初筛（可设0）
    df_s = sc.var_filter(df_train, y="SeriousDlqin2yrs", iv_limit=0.02, missing_limit=0.98, identical_limit=0.98)
    print("df_s.shape: ", df_s.shape)
    
    # 基于tree的分箱
    bins = sc.woebin(df_s, method='tree', count_distr_limit=0.04, bin_num_limit=5, y="SeriousDlqin2yrs")
    print("bins: ", bins['age'])
    
    # 分箱调整
    breaks_adj = {'age': [35, 55, 65]}
    bins_adj = sc.woebin(df_s, breaks_list=breaks_adj, method='tree', count_distr_limit=0.04, bin_num_limit=5, y="SeriousDlqin2yrs")
    print("bins_adj: ", bins_adj['age'])
    
    # iv值筛选
    fea_delete_iv = list({k: v['total_iv'][0] for k, v in bins_adj.items() if v['total_iv'][0] <= 0.1}.keys())
    for i in fea_delete_iv:
        del bins_adj[i]
    fea_after_iv = list(bins_adj.keys())
    print("fea_after_iv: ", fea_after_iv)
    
    fea_y = fea_after_iv
    fea_y.append('SeriousDlqin2yrs')
    print("fea_y: ", fea_y)
    
    # 相关性筛选
    corr_selected, dropped = toad.selection.select(df_train[fea_y], target='SeriousDlqin2yrs', empty=0, iv=0, corr=0.7, return_drop=True)
    fea_selected = list(corr_selected.columns)
    
    # woe transfer
    train_woe = sc.woebin_ply(df_train[fea_selected], bins_adj)
    test_woe = sc.woebin_ply(df_test[fea_selected], bins_adj)
    
    # 特征psi检验
    print(sc.perf_psi(
        score={'train': pd.DataFrame(round(train_woe['age_woe'], 4)), 'test': pd.DataFrame(round(test_woe['age_woe'], 4))},
        label={'train': train_woe['SeriousDlqin2yrs'], 'test': test_woe['SeriousDlqin2yrs']},
        show_plot=False
    ))
    
    print(sc.perf_psi(
        score={'train': pd.DataFrame(round(train_woe['RevolvingUtilizationOfUnsecuredLines_woe'], 4)),
               'test': pd.DataFrame(round(test_woe['RevolvingUtilizationOfUnsecuredLines_woe'], 4))},
        label={'train': train_woe['SeriousDlqin2yrs'], 'test': test_woe['SeriousDlqin2yrs']},
        show_plot=False
    ))
    
    print(sc.perf_psi(
        score={'train': pd.DataFrame(round(train_woe['NumberOfTimes90DaysLate_woe'], 4)),
               'test': pd.DataFrame(round(test_woe['NumberOfTimes90DaysLate_woe'], 4))},
        label={'train': train_woe['SeriousDlqin2yrs'], 'test': test_woe['SeriousDlqin2yrs']},
        show_plot=False
    ))
    
    # 模型训练
    y_train = train_woe.loc[:, 'SeriousDlqin2yrs']
    x_train = train_woe.loc[:, train_woe.columns != 'SeriousDlqin2yrs']
    y_test = test_woe.loc[:, 'SeriousDlqin2yrs']
    x_test = test_woe.loc[:, train_woe.columns != 'SeriousDlqin2yrs']
    
    lr = LogisticRegression(penalty='l2', C=0.9, solver='saga', n_jobs=-1)
    lr.fit(x_train, y_train)
    
    train_pred = lr.predict_proba(x_train)[:, 1]
    test_pred = lr.predict_proba(x_test)[:, 1]
    test_predict = lr.predict(x_test)
    
    print('perf_eva train: ', sc.perf_eva(y_train, train_pred, title="train", show_plot=False))
    print('perf_eva test: ', sc.perf_eva(y_test, test_pred, title="test", show_plot=False))
    
    # 评分卡转换
    card = sc.scorecard(bins_adj, lr, x_train.columns)
    
    # credit score
    train_score = sc.scorecard_ply(df_train, card, print_step=0)
    test_score = sc.scorecard_ply(df_test, card, print_step=0)
    
    # 转换后检验
    sc.perf_eva(y_train, np.array(train_score.iloc[:, 0]), title="train", show_plot=False)
    sc.perf_eva(y_test, np.array(test_score.iloc[:, 0]), title="test", show_plot=False)
    
    # 评分分布&psi
    sc.perf_psi(
        score={'train': train_score, 'test': test_score},
        label={'train': y_train, 'test': y_test},
        show_plot=False
    )
    
    # 提升度
    df_result = pd.DataFrame({"label": np.array(y_test), "predict": test_predict, "pos_probablity": test_pred,
                              "score": np.array(test_score.iloc[:, 0])})
    df_result['score_group'] = pd.qcut(df_result['score'], 8, duplicates='drop')
    
    # 保存df_result
    print("df_result: ", df_result.shape)
    
    df_result_bin = pd.crosstab(df_result['score_group'], df_result['label'])
    df_result_bin.insert(0, 'cut_number', range(0, len(df_result_bin)))  # 添加序列号 方便排序
    df_result_bin['total'] = df_result_bin[0] + df_result_bin[1]
    df_result_bin['cum_bad'] = df_result_bin[1].cumsum()
    df_result_bin['bad_by_model'] = df_result_bin[1] / df_result_bin[1].sum()
    df_result_bin['bad_by_random'] = df_result_bin['total'] / df_result_bin['total'].sum()  # random 分子分母同时乘以bad_rate 固定
    df_result_bin['cumbad_by_model'] = df_result_bin['cum_bad'] / df_result_bin[1].sum()
    df_result_bin['cumbad_by_random'] = df_result_bin['total'].cumsum() / df_result_bin['total'].sum()
    df_result_bin['lift'] = df_result_bin['cumbad_by_model'] / df_result_bin['cumbad_by_random']
    
    # 保存df_result_bin
    print("df_result_bin: ", df_result_bin)
    
    joblib.dump(card, './scorecard.pkl')

def saveFile(name, value):
    fieldnames = ['result_type', 'value']
    with open('./result.csv', 'w', newline='') as csvfile:
        writer = csv.DictWriter(csvfile, filednames=fieldnames)
        writer.writeheader()
        writer.writerow({'result_type': name, 'value': value})

def get_data(path):
    data = pd.read_csv(path)
    return data

data1 = get_data('./cs-training1.csv')
data2 = get_data('./cs-training2.csv')

train(data1, data2)
`

**Comment by zhouaihui** [comment]
这个代码看着似乎和secretflow没有关系，纯粹是想在occlum上面跑一段python代码。
一点建议：occlum里面访问host上的文件，需要使用路径 `/host/`开头，且文件必须放置在occlum_instance目录下。

**Comment by Jymjiaym** [comment]
> 这个代码看着似乎和secretflow没有关系，纯粹是想在occlum上面跑一段python代码。 一点建议：occlum里面访问host上的文件，需要使用路径 `/host/`开头，且文件必须放置在occlum_instance目录下。

目前这个代码是一个评分卡代码，在tee上跑不通，该代码能跑通后才能将alice和Bob参与进来，测试评分卡模型。
我们之前做的模型训练都是三方参与，将alice和bob文件发送给tee，tee做模型训练，这次也是类似的，没用到secretflow的api。
目前就是基本的python脚本，在occlum上引入评分卡模块和toad模块后就报以上错误了

**Comment by zhouaihui** [comment]
明白。那可以先修正一下路径试试。

**Comment by Jymjiaym** [comment]
> 明白。那可以先修正一下路径试试。

刚修正了 路径改了/host/文件，文件也挪到了/root/occlum_instance下，报错如下:
![7](https://github.com/secretflow/secretflow/assets/93430715/e8ae943f-706a-45f6-857b-0a1de3c9a019)
![8](https://github.com/secretflow/secretflow/assets/93430715/9596e9a2-5ed4-4493-9e4d-b057b81901bc)
